Tune the value of `min_marks_for_concurrent_read` to get better performance

jiebinn opened the pull request:
Tune the value of `min_marks_for_concurrent_read` on 80x2 vCPUs system. The default value is 24 and we have found if switch to 16. The overall geomean of Clickbench has got 4.3% performance gain. More than 10 queries have got 10% performance gain. For example, Q20 has got 12.6% performance more. By collecting the performance counters, the CPU utilization of Q20 has got 15% more with the patch, which would contribute to the QPS gain. Then there will be less idle CPU. From the pipeline figures, the pipeline of Q20 would be more compact.

After the patch
![image](https://github.com/user-attachments/assets/5210df50-63fd-4a79-87e0-63b4dbddef9d)

With the patch
![image](https://github.com/user-attachments/assets/dbb85ebf-3977-44e3-a9ce-bd9b62814519)


<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/jiebinsu/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/jiebinsu/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>
<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:11.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:Calibri, sans-serif;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
.xl65
	{border:.5pt solid windowtext;}
.xl66
	{mso-number-format:"0\.0%";
	border:.5pt solid windowtext;}
-->
</style>
</head>

<body link="#0563C1" vlink="#954F72">


Query | QPS gain
-- | --
0 | 102.5%
1 | 98.9%
2 | 98.2%
3 | 98.0%
4 | 99.1%
5 | 99.9%
6 | 94.5%
7 | 100.9%
8 | 111.2%
9 | 111.0%
10 | 111.9%
11 | 111.4%
12 | 106.5%
13 | 106.2%
14 | 106.7%
15 | 100.4%
16 | 101.0%
17 | 99.8%
18 | 104.2%
19 | 100.5%
20 | 112.6%
21 | 107.2%
22 | 106.3%
23 | 112.0%
24 | 113.9%
25 | 113.8%
26 | 113.6%
27 | 108.3%
28 | 105.5%
29 | 98.6%
30 | 107.6%
31 | 110.0%
32 | 101.8%
33 | 106.1%
34 | 107.5%
35 | 101.6%
36 | 100.4%
37 | 99.6%
38 | 100.7%
39 | 99.8%
40 | 100.0%
41 | 101.0%
42 | 100.5%
Geomean | 104.3%



</body>

</html>



<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Tune the value of `min_marks_for_concurrent_read` to get better performance.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ] <!---ci_set_required--> Allow: All Required Checks
- [ ] <!---ci_include_stateless--> Allow: Stateless tests
- [ ] <!---ci_include_stateful--> Allow: Stateful tests
- [ ] <!---ci_include_integration--> Allow: Integration Tests
- [ ] <!---ci_include_performance--> Allow: Performance tests
- [ ] <!---ci_set_builds--> Allow: All Builds
- [ ] <!---batch_0_1--> Allow: batch 1, 2 for multi-batch jobs
- [ ] <!---batch_2_3--> Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ] <!---ci_exclude_style--> Exclude: Style check
- [ ] <!---ci_exclude_fast--> Exclude: Fast test
- [ ] <!---ci_exclude_asan--> Exclude: All with ASAN
- [ ] <!---ci_exclude_tsan|msan|ubsan|coverage--> Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ] <!---ci_exclude_aarch64|release|debug--> Exclude: All with aarch64, release, debug
---
- [ ] <!---ci_include_fuzzer--> Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ] <!---ci_exclude_ast--> Exclude: AST fuzzers
---
- [ ] <!---do_not_test--> Do not test
- [ ] <!---woolen_wolfdog--> Woolen Wolfdog
- [ ] <!---upload_all--> Upload binaries for special builds
- [ ] <!---no_merge_commit--> Disable merge-commit
- [ ] <!---no_ci_cache--> Disable CI cache

jiebinn commented:
@nickitat , could you help to add the "can be tested" label?

robot-clickhouse-ci-1 commented:
<!-- automatic status comment for PR #69547 from jiebinn/ClickHouse:min_marks_for_concurrent_read -->
*This is an automated comment for commit a25b73329b4856a35adbe881b62c658f198e7dfc with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/69547/a25b73329b4856a35adbe881b62c658f198e7dfc/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/69432/6f9d1631e26bbb0263962c6d50c65e9ee2be561a/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

jiebinn commented:
There are some regressions while the query comes with `format Null`. I will investigate it after the National Holiday.

FrankChen021 commented:
I see this improvement is based on one specific machine which has larger number CPU cores. 
i'm wondering is it helpful in most scenarios, if not, it's better to keep this change on a local branch.

nickitat commented:
jfyi, currently i'm trying to get rid essentially of `merge_tree_min_*_for_concurrent_read_for_remote_filesystem` in this pr https://github.com/ClickHouse/ClickHouse/pull/69778

jiebinn commented:
Need to deal with some urgent task recently. I will track this PR ASAP.

alexey-milovidov commented:
@jiebinn Add changes to SettingsChangesHistory.

jiebinn commented:
Apologies for the delayed response. 
I have optimized the `min_marks_for_concurrent_read` parameter using Clickbench across various systems with 480, 256, and 64 vCPUs to achieve optimal performance. This tuning resulted in up to a 4.2% overall performance improvement for the geometric mean of 43 Clickbench queries on systems with 256 vCPUs. The adjustment shows significant performance gains on high core count systems, with no performance regression observed on standard systems.

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/jiebinsu/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/jiebinsu/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>
<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:11.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:Calibri, sans-serif;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
.xl63
	{border:.5pt solid windowtext;}
.xl64
	{mso-number-format:"0\.0%";
	border:.5pt solid windowtext;}
.xl65
	{text-align:center;
	border:.5pt solid windowtext;}
-->
</style>
</head>

<body link="#0563C1" vlink="#954F72">


cores | min_marks_for_concurrent_read | Geomean of   Clickbench
-- | -- | --
480   vCPUs | 12 | 100.7%
16 | 103.3%
20 | 101.3%
24 | 100.0%
28 | 99.3%
256   vCPUs | 4 | 100.7%
8 | 103.3%
12 | 102.1%
16 | 104.2%
20 | 103.4%
24 | 100.0%
28 | 98.7%
32 | 98.1%
64   vCPUs | 12 | 98.7%
16 | 100.3%
20 | 100.1%
24 | 100.0%
28 | 99.2%



</body>

</html>

jiebinn committed:
This patch have optimized the min_marks_for_concurrent_read parameter using
Clickbench across various systems with 480, 256, and 64 vCPUs to achieve
optimal performance. This tuning resulted in up to a 4.2% overall performance
improvement for the geometric mean of 43 Clickbench queries on systems with
256 vCPUs. The adjustment shows significant performance gains on high core
count systems, with no performance regression observed on standard systems.

Signed-off-by: Jiebin Sun <jiebin.sun@intel.com>

jiebinn committed:
Merge branch 'master' into min_marks_for_concurrent_read

jiebinn committed:
Merge branch 'ClickHouse:master' into min_marks_for_concurrent_read

jiebinn committed:
Merge branch 'master' into min_marks_for_concurrent_read

jiebinn committed:
Merge branch 'master' into min_marks_for_concurrent_read

jiebinn committed:
Merge branch 'master' into min_marks_for_concurrent_read

jiebinn commented:
It seems there is some regression during the CI test. I will close the PR first and think about that.
